utils/managed_bytes.hh

```diff
@@ -68,34 +88,117 @@ struct blob_storage {
+
+    single_chunk_blob_storage(single_chunk_blob_storage&& o) noexcept
+        : backref(o.backref)
+    {
+        backref->ptr = this;
+        memcpy(data, o.data, backref->size);
+    }
+
+    size_t storage_size() const noexcept {
+        return sizeof(*this) + backref->size;
+    }
+};
+
 // A managed version of "bytes" (can be used with LSA).
+//
+// Sometimes also used as a general-purpose fragmented buffer outside of LSA context,
+// but this is not recommended, because it's too easy to accidentally destroy it
+// in a different allocator than it was allocated in, which can break the program
+// in a hard-to-predict way.
+//
+// managed_bytes has three storage layouts:
+// 1. Inline.
+//    Used for data which fits into max_inline_size.
+// 2. External contiguous. (Single-allocation).
+//    Used for data which fits into preferred_max_contiguous_allocation().
+//    (At the moment of writing: 128 kiB and 12.8 kiB in LSA).
+//    The storage is a single single_chunk_blob_storage object.
+// 3. External fragmented. (Multi-allocation).
+//    Used for everything else.
+//    The storage is a chain of multi_chunk_blob_storage objects.
+//
+// Layout 2 exists as an optimization for the most common allocation sizes (several bytes).
+// There is nothing which prevents implementing these with layout 3, but layout 3 stores slightly
+// more metadata in the allocated buffer (pointer to the next fragment, size of the current fragment),
+// which adds up to a big overhead when used with small allocations.
+// E.g. 8-byte values are allocated externally -- each has additional 1 byte of flags and 8 bytes
+// of timestamp, so it's 17 bytes in total and that doesn't fit into inline storage.
+// And adding 16 bytes to each 17-byte cell is a big waste.
+//
+// The code of `class managed_bytes` is responsible for allocating and freeing the storage.
+// Code responsible for reading and writing it is in managed_bytes_basic_view.
+// The implementation details of these two classes are entangled.
 class managed_bytes {
     friend class bytes_ostream;
     static constexpr size_t max_inline_size = 15;
-    struct small_blob {
-        bytes_view::value_type data[max_inline_size];
-        int8_t size; // -1 -> use blob_storage
-    };
+
+    // The current layout is discerned by `inline_size`:
+    // >0 -> layout 1 (inline). In this case, the value of `inline_size` holds the data size.
+    // -1 -> layout 2 (single_chunk_blob_storage)
+    // -2 -> layout 3 (multi_chunk_blob_storage)
+    int8_t _inline_size = 0;
     union u {
         u() {}
         ~u() {}
-        blob_storage::ref_type ptr;
-        small_blob small;
+        bytes_view::value_type inline_data[max_inline_size]; // Stores the data directly. Size is in inline_size.
+        single_chunk_blob_storage::ref_type single_chunk_ref; // Points to external storage and stores the data size.
+        multi_chunk_blob_storage::ref_type multi_chunk_ref; // Points to external storage.
     } _u;
```

Comment:
I clang able to optimize moves to a couple of instructions (merging the copying of _inline_size and _u)? Please verify manually.

Please add perf-simple-query report to the cover letter, both read and write, this is critical infrastructure.

Reply:
Alright.

Reply:
Still missing. Also, please rebase to destale the test results.

Reply:
> Still missing. Also, please rebase to destale the test results.

Rebased and added the test results to the cover letter.

> I clang able to optimize moves to a couple of instructions (merging the copying of _inline_size and _u)? Please verify manually.

No, in fact the compiler insist on copying `_inline_size` and `_u` separately, even though their loads and stores are begging to be merged. I don't know why.

I put a manual memcpy in the move constructor to fix that. (But does it really matter?)